### PR TITLE
gitserver: Only support name/.git paths

### PR DESCRIPTION
**Pull request by keegancsmith**
@sourcegraph/core-services This doesn't update all the tests that break outside of the gitserver tests (mostly tests that use our gittest package that makes incorrect assumptions). I wanted feedback on if this approach is good before updating them.

Originally gitserver cloned bare repos to `ReposDir/name`. However, this
didn't allow any child repos like `ReposDir/name/subname` (which some
customers on gitolite have). So we migrated all repos to be stored at
`ReposDir/name/.git`. To prevent disruption we supported reading repositories
at both the old style and new style. The old style should not exist
anymore (the migration is so old, we have removed it). We didn't remove the
backwards compatibility support due to fear of breaking something. This commit
removes that support.

To remove the support we introduce a tiny type `GitDir`, and all functions
which operate on git directories now take a GitDir instead of a string. This
made it possible to audit all the call sites and how they interpreted the
directory. Now GitDir will always point to a `.git`, which simplifies many
assumptions going forward.

Fixes https://github.com/sourcegraph/sourcegraph/issues/2496

### CODE REVIEW

File: cmd/gitserver/server/server.go [hunk 1]
```diff
@@ -333,11 +333,15 @@ func (s *Server) handleIsRepoCloneable(w http.ResponseWriter, r *http.Request) {
 		http.Error(w, err.Error(), http.StatusBadRequest)
 		return
 	}
-	req.Repo = protocol.NormalizeRepo(req.Repo)
 
 	if req.URL == "" {
+		if req.Repo == "" {
+			http.Error(w, "no URL or Repo", http.StatusBadRequest)
```
Comment (unknwon) [inline cmd/gitserver/server/server.go]: nit: 
```suggestion
			http.Error(w, "no URL and Repo", http.StatusBadRequest)
```

The logic seems to have both ` == ""`.

File: cmd/gitserver/server/serverutil.go [hunk 2]
```diff
@@ -17,9 +17,40 @@ import (
...
+	"github.com/sourcegraph/sourcegraph/internal/api"
+	"github.com/sourcegraph/sourcegraph/internal/gitserver/protocol"
 	log15 "gopkg.in/inconshreveable/log15.v2"
 )
 
+// GitDir is an absolute path to a GIT_DIR.
+// They will all follow the form:
+//
+//    ${s.ReposDir}/${name}/.git
+type GitDir string
```
Comment (unknwon) [inline cmd/gitserver/server/serverutil.go]: I like this approach, but seems no way to enforce "the form" (`${s.ReposDir}/${name}/.git`) if some code assigns an inappropriate value (i.e. not ending with "/.git").

Comment (keegancsmith) [reply cmd/gitserver/server/serverutil.go]: We "enforce" this by all creations of `GitDir` going through the `dir` function. There are two places (outside of tests) where we don't do this, but they have comments saying why it is valid. We could make it impossible to bypass, but I think that is not really worth it vs the simplicity of this solution.

**Comment by keegancsmith** [comment]
@sourcegraph/core-services what do you think of this approach? Worth finishing off?

**Pull request merged**